volatility3/framework/plugins/windows/mftscan.py

```diff
@@ -43,104 +69,213 @@ def _generator(self):
+            attr_data = attr.Attr_Data.cast(si_object)
+            yield 0, (
+                format_hints.Hex(attr_data.vol.offset),
+                mft_record.get_signature(),
+                mft_record.RecordNumber,
+                mft_record.LinkCount,
+                mft_flag,
+                renderers.NotApplicableValue(),
+                attr.Attr_Header.AttrType.lookup(),
+                conversion.wintime_to_datetime(attr_data.CreationTime),
+                conversion.wintime_to_datetime(attr_data.ModifiedTime),
+                conversion.wintime_to_datetime(attr_data.UpdatedTime),
+                conversion.wintime_to_datetime(attr_data.AccessedTime),
+                renderers.NotApplicableValue(),
+            )
+
+        # File Name Attribute
+        elif attr.Attr_Header.AttrType.lookup() == "FILE_NAME":
+            fn_object = symbol_table + constants.BANG + "FILE_NAME_ENTRY"
+
+            attr_data = attr.Attr_Data.cast(fn_object)
+            file_name = attr_data.get_full_name()
+
+            # If we don't have a valid enum, coerce to hex so we can keep the record
+            try:
+                permissions = attr_data.Flags.lookup()
+            except ValueError:
+                permissions = hex(attr_data.Flags)
+
+            yield 1, (
+                format_hints.Hex(attr_data.vol.offset),
+                mft_record.get_signature(),
+                mft_record.RecordNumber,
+                mft_record.LinkCount,
+                mft_flag,
+                permissions,
+                attr.Attr_Header.AttrType.lookup(),
+                conversion.wintime_to_datetime(attr_data.CreationTime),
+                conversion.wintime_to_datetime(attr_data.ModifiedTime),
+                conversion.wintime_to_datetime(attr_data.UpdatedTime),
+                conversion.wintime_to_datetime(attr_data.AccessedTime),
+                file_name,
+            )
+
+    @staticmethod
+    def parse_data_record(
+        mft_record: interfaces.objects.ObjectInterface,
+        attr: interfaces.objects.ObjectInterface,
+        record_map: Dict[int, Tuple[str, int, int]],
+        return_first_record: bool,
+    ) -> Generator[Iterable, None, None]:
+        """
+        Returns the parsed data from a MFT record
+        """
+        # we only care about resident data
+        if attr.Attr_Header.NonResidentFlag:
+            return
+
+        # we aren't looking ADS when we want the first data record
+        if return_first_record:
+            ads_name = renderers.NotApplicableValue()
+
+        # skip records without a name if we want ADS entries
+        elif attr.Attr_Header.NameLength == 0:
+            return
+
+        else:
+            # past the first $DATA record, attempt to get the ADS name
+            # NotAvailableValue = > 1st Data, but name was not parsable
+            ads_name = attr.get_resident_filename()
+            if not ads_name:
```

Comment:
Feels like this could be `... = attr.get_resident_filename() or renderers...`  but I'm fine if you want to keep it this way.

Reply:
done